### ISSUE TITLE
Ability to have an APP_BEGIN_TIME constant

### ISSUE THREAD

**Issue by pentium10** [simplesamlphp/simplesamlphp#2064]
**Describe the solution you'd like**
Have a constant having init time, when the SSP flow starts

**Additional context**
We need to measure the time taken for a part of a script for code coverage and histogram. 
For this we are looking to have a ```APP_BEGIN_TIME``` constant used as below. 
```
/**
     * Returns the total time for serving the current request.
     * This method calculates the difference between now and the timestamp
     * defined by constant APP_BEGIN_TIME.
     * To estimate the execution time more accurately, the constant should
     * be defined as early as possible (best at the beginning of the entry script.)
     * @return float the total time for serving the current request.
     */
    public function getExecutionTime() {
        return (function_exists('hrtime') ? hrtime(true) / 1e9 : microtime(true)) - APP_BEGIN_TIME;
    }
```

What are the current possiblities? Can we have this feature?

**Comment by tvdijen** [comment]
What do you mean with 'for code coverage and histogram'?  You want to make sure that code changes do not impact performance too much?

**Comment by pentium10** [comment]
exactly, more because of plug-ins, modules which could be third party, or external dependencies.

**Comment by tvdijen** [comment]
I understand, but execution time is always variable, that's why I ask this.. Wouldn't it make more sense for you to measure ticks? I'm not sure if you're familiar with the concept of ticks, but basically you're counting operations that are being executed. This gives you a _true_ understanding if your code-change is increasing/decreasing performance, rather than execution time which may vary because your OS is waiting for an interrupt to happen to catch a file from disk (for example)

**Comment by tvdijen** [comment]
I got myself familiar with ticks over two decades ago and it's a _great_ way to tell you what you're doing and what you can do to optimize your code...

**Comment by pentium10** [comment]
The industry around PHP frameworks have Loggers that want this information supplied. Its commonly used in different frameworks, a couple of examples: [LARAVEL_START](https://github.com/laravel/laravel/blob/cf0b40b878b5068ed710ee39a0e1d090a1f06d60/public/index.php#L5), or [YII_BEGIN_TIME](https://github.com/yiisoft/yii2/blob/8861f7bd9daf66cf7db7edaeefc236e71685273e/framework/BaseYii.php#L20). I understand the concept of ticks as well, but some roundtrips can be measured with request time.

**Comment by tvdijen** [comment]
Ok, works for me! I just wanted to understand the question. I see no harm in setting this variable, so i will include it in a next release

**Comment by tvdijen** [comment]
Does it matter for you if we use `hrtime` or `microtime`?  `hrtime` seems the better one of the two and is available for PHP 7+..

**Comment by pentium10** [comment]
hrtime is good, and it's recommended.

**Comment by tvdijen** [comment]
Closed in 9a58f6edace12d6032a5a369ce60a15864113b29
